modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/dht/GridCachePartitionsStateValidationTest.java

```diff
@@ -323,4 +330,139 @@ public void waitUntilAllSingleMessagesAreSent() throws IgniteCheckedException {
+        testLogger = new ListeningTestLogger();
+
+        LogListener lsnr = LogListener.matches("Partitions update counters are inconsistent for ").build();
+
+        testLogger.registerListener(lsnr);
+
+        try {
+            IgniteEx ignite = startGrid(0);
+            startGrid(1);
+            startGrid(2);
+            ignite.cluster().active(true);
+            awaitPartitionMapExchange();
+
+            // Populate cache to increment update counters.
+            for (int i = 0; i < 1000; i++)
+                ignite.cache(CACHE_NAME).put(i, i);
+
+            // Modify update counter for some partition.
+            for (GridDhtLocalPartition partition : ignite.cachex(CACHE_NAME).context().topology().localPartitions()) {
+                partition.updateCounter(100500L);
+                break;
+            }
+
+            // Trigger exchange.
+            startGrid(3);
+
+            awaitPartitionMapExchange();
+
+            // Nothing should happen (just log error message) and we're still able to put data to corrupted cache.
+            ignite.cache(CACHE_NAME).put(0, 0);
+
+
+            assertTrue("Counters inconsistent message not found", lsnr.check());
+        }
+        finally {
+            stopAllGrids();
+        }
+    }
+
+    @Test
+    public void testValidationIfPartitionSizesAreInconsistentWithSize() throws Exception {
+
+        testLogger = new ListeningTestLogger();
+
+        LogListener lsnrSize = LogListener.matches("Partitions cache sizes are inconsistent").build();
+        testLogger.registerListener(lsnrSize);
+
+        try{
+            IgniteEx ignite = startGrid(0);
+            startGrid(1);
+            startGrid(2);
+            ignite.cluster().active(true);
+            awaitPartitionMapExchange();
+
+            // Populate cache to increment update counters.
+            for (int i = 0; i < 1000; i++)
+                ignite.cache(CACHE_NAME).put(i, i);
+
+            // Modify update size for some partition.
+            for (GridDhtLocalPartition partition : ignite.cachex(CACHE_NAME).context().topology().localPartitions()) {
+                partition.dataStore().clear(ignite.cachex(CACHE_NAME).context().cacheId());
+                break;
+            }
+
+            // Trigger exchange.
+            startGrid(3);
+
+            awaitPartitionMapExchange();
+
+            // Nothing should happen (just log error message) and we're still able to put data to corrupted cache.
+            ignite.cache(CACHE_NAME).put(0, 0);
+
+            assertTrue("Size inconsistent message not found",lsnrSize.check());
+        }
+        finally {
+            stopAllGrids();
+        }
+    }
+
+
+    @Test
+    public void testValidationIfPartitionSizesAndPartitionCountersAreInconsistentWithSize() throws Exception {
+
+        testLogger = new ListeningTestLogger();
+
+        LogListener lsnrSize = LogListener.matches("Partitions cache sizes are inconsistent").build();
```

Comment:
This test checks a multiline error message so I suggest to use single logListener for both parts of message. LogListener.matches supports regular expressions so it shouldn't be too difficult.